am-lib/src/main/java/uk/gov/hmcts/reform/amlib/internal/FilterService.java

```diff
@@ -19,39 +20,84 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
+import static java.util.Objects.isNull;
 import static uk.gov.hmcts.reform.amlib.enums.Permission.READ;
+import static uk.gov.hmcts.reform.amlib.enums.SecurityClassification.NONE;
 
 @Slf4j
 @SuppressWarnings("LineLength")
 public class FilterService {
 
     private static final JsonPointer WHOLE_RESOURCE_POINTER = JsonPointer.valueOf("");
 
-    public JsonNode filterJson(JsonNode resource, Map<JsonPointer, Set<Permission>> attributePermissions) {
-        List<JsonPointer> nodesWithRead = filterPointersWithReadPermission(attributePermissions);
-        log.debug("> Nodes with READ access: " + nodesWithRead);
+    public JsonNode filterJson(JsonNode resource,
+                               Map<JsonPointer, Set<Permission>> attributePermissions,
+                               Map<JsonPointer, SecurityClassification> attributeSecurityClassifications,
+                               Set<SecurityClassification> userSecurityClassifications) {
 
-        if (nodesWithRead.isEmpty()) {
-            return null;
-        }
+        List<JsonPointer> nodesToKeep = filterPointersWithReadPermission(attributePermissions);
+        log.debug("> Nodes with READ access: " + nodesToKeep);
 
         JsonNode resourceCopy = resource.deepCopy();
 
-        if (!nodesWithRead.contains(WHOLE_RESOURCE_POINTER)) {
-            List<JsonPointer> uniqueNodesWithRead = reducePointersToUniqueList(nodesWithRead);
+        List<JsonPointer> nodesToRemove = filterPointersWithoutReadPermission(attributePermissions);
```

Comment:
from performance point of view, it may be better to  combine the logic to filter nodes with read permissions and without into one

Reply:
👍